From d4dee89e97952dff63a65ea850ab7aeac2d1f9c0 Mon Sep 17 00:00:00 2001
Message-Id: <d4dee89e97952dff63a65ea850ab7aeac2d1f9c0.1513880568.git.jan.steffens@gmail.com>
From: "Jan Alexander Steffens (heftig)" <jan.steffens@gmail.com>
Date: Thu, 21 Dec 2017 18:22:55 +0100
Subject: [PATCH] shared/compat: Fix memory handling of
 nm_setting_vpn_get_*_keys (v2)

The compat implementations return a (transfer none) strv instead of a
(transfer container) one. This has caused double frees in nm-applet:
https://bugs.archlinux.org/task/56772

We still need to copy the keys because nm_setting_vpn_foreach_* provides
us with copies that are freed after the iteration.

Fix this by handing out a duplicate of the array.

Fixes: 272439cb2077a889485dd195bb6b99ca91d3e3bf
---
 shared/nm-utils/nm-compat.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/shared/nm-utils/nm-compat.c b/shared/nm-utils/nm-compat.c
index 22ab675de8373639..ba50a509a3a5f530 100644
--- a/shared/nm-utils/nm-compat.c
+++ b/shared/nm-utils/nm-compat.c
@@ -54,15 +54,15 @@ _get_keys (NMSettingVpn *setting,
 	if (a->len) {
 		g_ptr_array_sort (a, nm_strcmp_p);
 		g_ptr_array_add (a, NULL);
-		keys = (const char **) g_ptr_array_free (g_steal_pointer (&a), FALSE);
+		keys = g_memdup (a->pdata, a->len * sizeof *keys);
 
 		/* we need to cache the keys *somewhere*. */
 		g_object_set_qdata_full (G_OBJECT (setting),
 		                         is_secrets
 		                         ? NM_CACHED_QUARK ("libnm._nm_setting_vpn_get_secret_keys")
 		                         : NM_CACHED_QUARK ("libnm._nm_setting_vpn_get_data_keys"),
-		                         keys,
-		                         (GDestroyNotify) g_strfreev);
+		                         g_ptr_array_ref (a),
+		                         (GDestroyNotify) g_ptr_array_unref);
 	}
 
 	NM_SET_OUT (out_length, len);
-- 
2.15.1

